frontend/templates/product-list/sections/FilterableProductsSection/index.tsx

```diff
@@ -65,6 +66,31 @@ export function FilterableProductsSection({
       () => filterFalsyItems(hits.map(productPreviewFromAlgoliaHit)),
       [hits]
    );
+
+   const [prevProductIds, setPrevProductIds] = React.useState<string[]>([]);
+
+   React.useEffect(() => {
+      const currentProductIds = products.map((product) => product.id);
+      if (
+         JSON.stringify(currentProductIds) !== JSON.stringify(prevProductIds)
+      ) {
+         if (currentProductIds.length > 0) {
+            debouncedTrackGA4ViewItemList({
+               item_list_id: productList.handle,
+               item_list_name: productList.title,
+               items: products.map((product) => ({
+                  item_id: product.sku,
+                  item_name: product.title,
+                  item_variant: product.id,
+                  quantity: product.quantityAvailable,
+                  price: product.price.amount,
+               })),
+            });
+         }
+         setPrevProductIds(currentProductIds);
+      }
+   }, [prevProductIds, products, productList.handle, productList.title]);
```

Comment:
I originally tried something similar to your proposed solution, and unfortunately it doesn't work.

1.) the useEffect() thinks the products has changed even though the contents are identical. I assume this is because it is comparing references and not contents. ergo why i feel the need to `JSON.stringify`
2.) the `prevProductIds` is probably unnecessary, but a microoptimization for comparisons
3.) This is due to some sort of issue where as far as i can tell hits is changing causing rapid reloads before the setPrevProductIds can even be called, almost like this component is being rerendered in parrallel. I tried to look into this but didn't figure it out. I think that solving an issue with what is likely an algolia issue is outside the scope of an issue to setup analytics. Ergo the `debounce` solution, so we dont rapid fire a bunch of duplicate ga4 events on page load.

Reply:
> the useEffect() thinks the products has changed even though the contents are identical. I assume this is because it is comparing references and not contents. ergo why i feel the need to JSON.stringify

Since `products` is the result of `useMemo`, react will only think `products` has changed when `hits` has changed ([ref](https://react.dev/reference/react/useMemo#should-you-add-usememo-everywhere)).

That being said, what you said makes it sound like `hits` changes a few times as the page loads, which could cause the `useEffect` callback to be evaluated several times with different `products`. Is it possible that's what is happening?

Reply:
It could also be related to [strict mode in dev](https://react.dev/reference/react/useEffect#my-effect-runs-twice-when-the-component-mounts) like Sterling said earlier https://github.com/iFixit/react-commerce/pull/1973#discussion_r1323667279.

Reply:
>It could also be related to [strict mode in dev](https://react.dev/reference/react/useEffect#my-effect-runs-twice-when-the-component-mounts) like Sterling said earlier https://github.com/iFixit/react-commerce/pull/1973#discussion_r1323667279.

as far as i can tell, we don't have strict mode setup in our app at all

Reply:
> That being said, what you said makes it sound like hits changes a few times as the page loads, which could cause the useEffect callback to be evaluated several times with different products

If this is the case, I wonder if we can use the debounce method and not necessarily need the previous state 🤷🏻